File: Unique/Accounts/Services/AccountService.cs
using Unique.Accounts.Repositories;
using Unique.World;

namespace Unique.Accounts.Services
{
    public sealed class AccountService
    {
        private readonly AccountRepository repository = new AccountRepository();
        private readonly PasswordService passwordService = new PasswordService();
        private readonly AccountValidationService validationService = new AccountValidationService();

        public void Load()
        {
            repository.Load();
        }

        public bool IsValidName(string input) => validationService.IsValidName(input);

        public bool IsValidEmail(string input) => validationService.IsValidEmail(input);

        public bool EmailExists(string email) => repository.EmailExists(email);

        public bool SocialClubExists(string socialClubId) => repository.SocialClubExists(socialClubId);

        public Account GetById(int accountId) => repository.GetById(accountId);

        public Account GetByEmail(string email) => repository.GetByEmail(email);

        public Account GetBySocialClubId(string socialClubId) => repository.GetBySocialClubId(socialClubId);

        public bool VerifyPassword(Account account, string password) => passwordService.VerifyPassword(account, password);

        public Account Create(string firstName, string lastName, string email, string password, string socialClubName, string socialClubId)
        {
            if (repository.EmailExists(email))
                return null;

            if (repository.SocialClubExists(socialClubId))
                return null;

            string salt = passwordService.CreateSalt();
            string hash = passwordService.HashPassword(password, salt);

            SpawnPoint spawn = ServerSpawnService.GetSpawn();

            var account = new Account
            {
                FirstName = string.IsNullOrWhiteSpace(firstName) ? "Charakter" : NormalizeName(firstName),
                LastName = string.IsNullOrWhiteSpace(lastName) ? "Erstellen" : NormalizeName(lastName),
                Email = email.Trim(),
                SocialClubName = NormalizeOptional(socialClubName),
                SocialClubId = NormalizeOptional(socialClubId),
                PasswordSalt = salt,
                PasswordHash = hash,
                CharacterCreated = false,
                AdminLevel = 0,
                Health = 100,
                Armor = 0,
                Dimension = spawn.Dimension,
                PosX = spawn.X,
                PosY = spawn.Y,
                PosZ = spawn.Z,
                RotZ = spawn.RotZ
            };

            repository.Add(account);
            return account;
        }

        public bool IsSocialClubMatch(Account account, string socialClubId)
        {
            if (account == null)
                return false;

            if (string.IsNullOrWhiteSpace(account.SocialClubId))
                return true;

            return string.Equals(account.SocialClubId, socialClubId?.Trim(), System.StringComparison.OrdinalIgnoreCase);
        }

        public void BindSocialClub(Account account, string socialClubName, string socialClubId)
        {
            if (account == null || string.IsNullOrWhiteSpace(socialClubId))
                return;

            account.SocialClubId = socialClubId.Trim();
            account.SocialClubName = NormalizeOptional(socialClubName);
            repository.SaveChanges(account);
        }

        public bool SetAdminLevel(int accountId, int adminLevel)
        {
            if (adminLevel < 0 || adminLevel > 10)
                return false;

            Account account = repository.GetById(accountId);
            if (account == null)
                return false;

            account.AdminLevel = adminLevel;
            repository.SaveChanges(account);
            return true;
        }

        public bool SavePosition(int accountId, float posX, float posY, float posZ, float rotZ)
        {
            Account account = repository.GetById(accountId);
            if (account == null)
                return false;

            account.PosX = posX;
            account.PosY = posY;
            account.PosZ = posZ;
            account.RotZ = rotZ;
            repository.SaveChanges(account);
            return true;
        }

        public bool SavePlayerState(int accountId, float posX, float posY, float posZ, float rotZ, uint dimension, int health, int armor, int cash, int bankCash)
        {
            Account account = repository.GetById(accountId);
            if (account == null)
                return false;

            account.PosX = posX;
            account.PosY = posY;
            account.PosZ = posZ;
            account.RotZ = rotZ;
            account.Dimension = dimension;
            account.Health = Clamp(health, 0, 100);
            account.Armor = Clamp(armor, 0, 100);
            account.Cash = ClampMoney(cash);
            account.BankCash = ClampMoney(bankCash);
            repository.SaveChanges(account);
            return true;
        }

        public bool SetCash(int accountId, int cash)
        {
            Account account = repository.GetById(accountId);
            if (account == null)
                return false;

            account.Cash = ClampMoney(cash);
            repository.SaveChanges(account);
            return true;
        }

        public bool SetBankCash(int accountId, int bankCash)
        {
            Account account = repository.GetById(accountId);
            if (account == null)
                return false;

            account.BankCash = ClampMoney(bankCash);
            repository.SaveChanges(account);
            return true;
        }

        public bool SetBanState(int accountId, bool isBanned, string reason)
        {
            Account account = repository.GetById(accountId);
            if (account == null)
                return false;

            account.IsBanned = isBanned;
            account.BanReason = isBanned ? NormalizeOptional(reason) ?? "Kein Grund angegeben." : null;
            repository.SaveChanges(account);
            return true;
        }

        public bool CompleteCharacter(int accountId, string firstName, string lastName, string birthDate, string origin, string customizationJson)
        {
            if (!validationService.IsValidName(firstName) || !validationService.IsValidName(lastName))
                return false;

            Account account = repository.GetById(accountId);
            if (account == null)
                return false;

            account.FirstName = NormalizeName(firstName);
            account.LastName = NormalizeName(lastName);
            account.BirthDate = NormalizeOptional(birthDate);
            account.Origin = NormalizeOptional(origin);
            account.CustomizationJson = NormalizeOptional(customizationJson);
            account.CharacterCreated = true;
            repository.SaveChanges(account);
            return true;
        }

        private static string NormalizeName(string input)
        {
            string value = input.Trim().ToLower();
            return char.ToUpper(value[0]) + value.Substring(1);
        }

        private static string NormalizeOptional(string input)
        {
            return string.IsNullOrWhiteSpace(input) ? null : input.Trim();
        }

        private static int Clamp(int value, int min, int max)
        {
            if (value < min)
                return min;

            return value > max ? max : value;
        }

        private static int ClampMoney(int value)
        {
            return value < 0 ? 0 : value;
        }
    }
}

File: server/src/features/accounts/account-service.ts
import { createHash, randomBytes, timingSafeEqual } from "crypto";
import { clamp, clampMoney } from "../../runtime/helpers.js";
import type { Account } from "./account.js";
import { AccountRepository } from "./account-repository.js";
import { SpawnService } from "../world/spawn-service.js";
import type { CompleteCharacterDto, CreateAccountDto, SavePlayerStateDto } from "./api/account-dtos.js";
import { normalizeOptionalString, normalizeName } from "../../shared/normalize.js";

function hashPassword(password: string, salt: string) {
  return createHash("sha256").update(`${salt}:${password}`).digest("hex");
}

function parseCustomizationJson(raw: string | null) {
  if (!raw) {
    return {};
  }

  try {
    const parsed = JSON.parse(raw);
    return parsed && typeof parsed === "object" ? parsed : {};
  } catch (error) {
    return {};
  }
}

import { PhoneService } from "../phone/phone-service.js";

export class AccountService {
  private readonly repository = new AccountRepository();
  private readonly spawns = new SpawnService();
  private readonly phoneService?: PhoneService;

  constructor(phoneService?: PhoneService) {
      this.phoneService = phoneService;
  }

  isValidName(input?: string | null) {
    return /^[A-Za-z-]{2,24}$/.test(input?.trim() ?? "");
  }

  isValidEmail(input?: string | null) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(input?.trim() ?? "");
  }

  async emailExists(email: string) {
    return this.repository.emailExists(email);
  }

  async socialClubExists(socialClubId: string) {
    return this.repository.socialClubExists(socialClubId);
  }

  async getById(accountId: number) {
    return this.repository.getById(accountId);
  }

  async getByEmail(email: string) {
    return this.repository.getByEmail(email);
  }

  async getBySocialClubId(socialClubId: string) {
    return this.repository.getBySocialClubId(socialClubId);
  }

  async getPersonalOutfits(accountId: number) {
    return this.repository.getPersonalOutfits(accountId);
  }

  async getPersonalOutfitById(outfitId: number, accountId: number) {
    return this.repository.getPersonalOutfitById(outfitId, accountId);
  }

  async createPersonalOutfit(accountId: number, name: string, clothingJson: string) {
    const count = await this.repository.countPersonalOutfits(accountId);
    if (count >= 20) return { ok: false as const, message: "Outfit-Limit (20) erreicht." };
    const outfit = await this.repository.createPersonalOutfit(accountId, name, clothingJson);
    return { ok: true as const, outfit };
  }

  async deletePersonalOutfit(outfitId: number, accountId: number) {
    return this.repository.deletePersonalOutfit(outfitId, accountId);
  }

  async create(dto: CreateAccountDto) {
    const spawn = await this.spawns.getSpawn();
    const salt = randomBytes(16).toString("hex");

    return this.repository.create({
      firstName: "Charakter",
      lastName: "Erstellen",
      email: dto.email.trim(),
      phoneNumber: null,
      socialClubName: normalizeOptionalString(dto.socialClubName),
      socialClubId: normalizeOptionalString(dto.socialClubId),
      passwordHash: hashPassword(dto.password, salt),
      passwordSalt: salt,
      characterCreated: false,
      birthDate: null,
      origin: null,
      customizationJson: null,
      adminLevel: 0,
      cash: 0,
      bankCash: 0,
      health: 100,
      armor: 0,
      dimension: spawn.dimension,
      posX: spawn.x,
      posY: spawn.y,
      posZ: spawn.z,
      rotZ: spawn.rotZ,
      isBanned: false,
      banReason: null,
      banDate: null,
      banExpiresAt: null,
      banAdminName: null,
      banAdminAccountId: 0
    });
  }

  verifyPassword(account: Account, password: string) {
    const incomingHash = hashPassword(password, account.passwordSalt);
    return timingSafeEqual(Buffer.from(incomingHash, "hex"), Buffer.from(account.passwordHash, "hex"));
  }

  isSocialClubMatch(account: Account, socialClubId?: string | null) {
    if (!account.socialClubId || !socialClubId) {
      return true;
    }

    return account.socialClubId === socialClubId;
  }

  async bindSocialClub(account: Account, name?: string | null, socialClubId?: string | null) {
    if (account.socialClubId || !socialClubId) {
      return account;
    }

    return this.repository.updateSocialClub(account.accountId, normalizeOptionalString(name), socialClubId);
  }

  async setBanState(accountId: number, banned: boolean, reason: string | null, expiresAt?: string | null, adminName?: string | null, adminAccountId?: number | null) {
    return this.repository.updateBanState(accountId, banned, reason, expiresAt, adminName, adminAccountId);
  }

  async completeCharacter(accountId: number, dto: CompleteCharacterDto) {
    return this.repository.updateCharacter(accountId, {
      firstName: normalizeName(dto.firstName),
      lastName: normalizeName(dto.lastName),
      characterCreated: true,
      birthDate: dto.birthDate,
      origin: dto.origin,
      customizationJson: dto.customizationJson
    });
  }

  async getByName(firstName: string, lastName: string) {
    return this.repository.getByName(firstName, lastName);
  }

  async savePlayerState(accountId: number, state: SavePlayerStateDto) {
    const customizationJson = normalizeOptionalString(state.customizationJson);
    const result = await this.repository.updateState(accountId, {
      cash: clampMoney(state.cash),
      bankCash: clampMoney(state.bankCash),
      health: clamp(state.health ?? 100, 0, 100),
      armor: clamp(state.armor ?? 0, 0, 100),
      dimension: state.dimension,
      posX: state.posX,
      posY: state.posY,
      posZ: state.posZ,
      rotZ: state.rotZ,
      customizationJson: state.customizationJson
    });

    // --- NOTIFICATION INTEGRATION ---
    if (this.phoneService && state.bankCash !== undefined) {
      const player = (mp.players as any).toArray().find((p: any) => Number(p.getVariable("ACCOUNT_ID")) === accountId);
      if (player) {
          const oldBank = Number(player.getVariable("BANK_CASH") ?? 0);
          if (state.bankCash !== oldBank) {
              const diff = state.bankCash - oldBank;
              this.phoneService.sendNotification(player, {
                  title: "Bank-Benachrichtigung",
                  content: `${diff > 0 ? '+' : ''}${diff}$ Gutschrift/Abbuchung. Neuer Kontostand: ${state.bankCash}$`,
                  icon: "fas fa-university",
                  app: "Wallet"
              });
          }
      }
    }
    // --- END NOTIFICATION INTEGRATION ---

    return result;
  }

  // Alias für Rückwärtskompatibilität während des Build-Übergangs
  async updateState(accountId: number, state: SavePlayerStateDto) {
    return this.savePlayerState(accountId, state);
  }
}
